Automattic/sensei
# Add course results block

Fixes #4249

### Changes proposed in this Pull Request

* This adds the basic parts of the course results block behind the established feature flag (`course_completed_page`).
* Since all color settings need to be universal (this block isn't per-course), I tried to make them generic and put the in the block sidebar.
* It tries to follow a lot of what we did for the course outline block for frontend rendering. I diverged a tiny bit for the editor (using CSS variables) just because it simplified the code now that everything is one block vs outline -> lesson/module blocks.
* Note: This looks a bit funny on courses that don't have quizzes, especially not within modules where you could add a border (see screenshot). We might need to iterate a bit on the design for those. However, it'd be nice to get this in before I go on leave so it might need to happen in a new PR.

### Partially implemented
Although implemented in the editor, I did not implement the overall course grade. This PR was getting quite big and is something that might need some more thought (for example: you can complete a course before the last quiz is graded if not auto-graded and required, how does that factor into it?). If we don't get to it, it will be easy to remove from the editor preview.

### Testing instructions
* If you haven't already, add `define( 'SENSEI_FEATURE_FLAG_COURSE_COMPLETED_PAGE', true );` to `wp-config.php`.
* Create a completed course page with this block. Make sure the Completed Course setting is set to the new page.
* Have a course that has a mix of lessons and modules. Let some have auto-graded quizzes.
* Complete the course.
* Verify you are redirected to the completed course page with the `?course_id=ID` param and the structure appears correctly.
* Modify the styles of the block and refresh the completed course page. Ensure the editor and frontend match.

### Screenshots
<details>
  <summary>Click to expand and see screenshots!</summary>

<img width="1354" alt="Screen Shot 2021-07-29 at 5 01 25 PM" src="https://user-images.githubusercontent.com/68693/127525904-949a6e57-0af6-47c7-82a5-b7722145e6bd.png">

<img width="836" alt="Screen Shot 2021-07-29 at 4 51 36 PM" src="https://user-images.githubusercontent.com/68693/127524223-1e1171f3-273e-4b3d-a8dd-809a61bb193a.png">

<img width="806" alt="Screen Shot 2021-07-29 at 4 52 38 PM" src="https://user-images.githubusercontent.com/68693/127524401-101bc93c-5549-4c3a-99e9-7bfe1b6c187d.png">

<img width="829" alt="Screen Shot 2021-07-29 at 4 53 14 PM" src="https://user-images.githubusercontent.com/68693/127524544-b00c8ff0-b04e-44be-af6a-3fc67e2ac2a9.png">
<img width="845" alt="Screen Shot 2021-07-29 at 4 56 26 PM" src="https://user-images.githubusercontent.com/68693/127525015-cd8e9e6d-84d5-42db-8daa-6b2d7045ea53.png">
</details>
